base/weakkeydict.jl

```diff
@@ -1,22 +1,78 @@
 # weak key dictionaries
 
 type WeakKeyDict{K,V} <: Associative{K,V}
-    ht::Dict{Any,V}
+    ht::Dict{WeakRef,V}
     lock::Threads.RecursiveSpinLock
     finalizer::Function
 
+    # Constructors mirror Dict's
     function WeakKeyDict()
         t = new(Dict{Any,V}(), Threads.RecursiveSpinLock(), identity)
-        t.finalizer = function(k)
+        t.finalizer = function (k)
             # when a weak key is finalized, remove from dictionary if it is still there
             islocked(t) && return finalizer(k, t.finalizer)
             delete!(t, k)
         end
         return t
     end
+    function WeakKeyDict(kv)
+        h = WeakKeyDict{K,V}()
+        for (k,v) in kv
+            h[k] = v
+        end
+        return h
+    end
+    WeakKeyDict(p::Pair) = setindex!(WeakKeyDict{K,V}(), p.second, p.first)
+    function WeakKeyDict(ps::Pair...)
+        h = WeakKeyDict{K,V}()
+        sizehint!(h, length(ps))
+        for p in ps
+            h[p.first] = p.second
+        end
+        return h
+    end
 end
 WeakKeyDict() = WeakKeyDict{Any,Any}()
 
+WeakKeyDict(kv::Tuple{}) = WeakKeyDict()
```

Comment:
These outer constructors are copies of the Dict ones.  It could be DRYed up using something along the lines of
```
type A1{T}
    a::T
    A1(x) = new(x)
end
type A2{T}
    a::T
    A2(x) = new(x)
end
typealias A Union{A1,A2}
(::Type{TA}){TA<:A,T}(x::T) = TA{T}(x)
```
However, I suspect that there will be boot-strap isssues as WeakKeyDict comes later, thus these constructors would need to be added to weakkeydict.jl.  Alternatively, I could put a comment into dict.jl to state to keep the constructors synchronized.